snakemake/dag.py

```diff
@@ -783,29 +777,35 @@ def update(
                     if file
                     else "",
                 )
-        if producer is None:
+        if not producers:
             if cycles:
                 job = cycles[0]
                 raise CyclicGraphException(job.rule, file, rule=job.rule)
             if len(exceptions) > 1:
                 raise WorkflowError(*exceptions)
             elif len(exceptions) == 1:
                 raise exceptions[0]
-        else:
-            logger.dag_debug(dict(status="selected", job=producer))
-            logger.dag_debug(
-                dict(
-                    file=file,
-                    msg="Producer found, hence exceptions are ignored.",
-                    exception=WorkflowError(*exceptions),
-                )
-            )
 
         n = len(self.dependencies)
         if progress and n % 1000 == 0 and n and self._progress != n:
             logger.info("Processed {} potential jobs.".format(n))
             self._progress = n
 
+        producers.sort(reverse=True)
```

Comment:
So far I could follow, but can't we run into the described sorting issue here again?

Reply:
No, we cannot run into the same problem.
What we actually need is the producer that strictly dominates other possible producers. That means that it shall be comparable to any other producer and shall be larger than each one. If such rule exists, it will get to the first position in the list (line 795). If there is no such rule, then we have an ambiguity (actually we may have multiple ambiguous rules). The lines 796-798 are needed just to produce a meaningful error message (not to complain on two rules that are not actually ambiguous).

Reply:
Even simpler would be to describe the issue in different words. Current code sorts first and then eliminates non-producing rules. My solution eliminates non-producing first and sorts after that. If the first case the order is not predictable and can be broken by "spoilers": the rules that would be eliminated later. In my solution the spoilers are removed first, and then sorting gives us some guarantees. First guarantee: if the unambiguous rule exists, it is on the first position. Second guarantee: whenever there is an ambiguity, the item on the first position shall be non-compairable with at least one rule on another position.

Reply:
Ah, right, that makes sense.